Port-Models doc (#321)
stanford-crfm/levanter: docs/Add-Models.md

> # Add New Models to Levanter

Reviewer: maybe let's call it Porting Models to Levanter?

This guide outlines the process of adding new models into Levanter. While we emphasize adding models that are implemented in Hugging Face, these steps are applicable to models from other sources.

We'll start with a detailed walkthrough on implementing your model and testing it. Subsequently, we'll describe how to configure and run a training job with your model. To conclude, we'll share insights and recommendations to enhance your model's training efficiency.

## Write Your Model

> Writing a new model in Levanter is very similar to Hugging Face. You start by adding a new file in [models/](https://github.com/stanford-crfm/levanter/tree/main/src/levanter/models). You first create a config class to register the key hyperparameters and axes of your model. Then, you write the model class that includes key layers and components of your model.

Reviewer: Maybe 'At a high level,' ? Writing with Haliax is pretty different...

### Write Config

> We start by writing your model config class. This class will register all the hyperparameters and axes of your models. We want to define them as the first step because they will be used immidately in the next step that implements your model.

Reviewer: Somewhere we should mention registration with LmConfig for LM Models. We should also mention that configs should be a dataclass to work with the config parsing stuff.

Maybe we should just provide a template?

```python
@LmConfig.register_subclass("gpt2") # if implementing a Causal Language model
@dataclass(frozen=True)
class MyConfig(HFCompatConfig):  # Extend HFCompatConfig if you want to support HF import/export
```

Note that you don't need to write all of configurations at once. You can start with the key hyperparameters and axes, and add more as you implement the model.

**Hyperparameters** should be declared with their type and a default value, illustrated below:

```python
seq_len: int = 2048
hidden_dim: int = 4096
intermediate_dim: int = 11008
```

> **Model axies** are used for parallelization. An Axis is registered with its name and size. The size of an Axis is normally associated with a hyperparameter. For example:

Reviewer: ```suggestion
**Model axes** are used for parallelization. An Axis is registered with its name and size. The size of an Axis is normally associated with a hyperparameter. For example:
```

```python
Pos = property(lambda self: Axis(name="position", size=self.seq_len))
Embed = property(lambda self: Axis(name="embed", size=self.hidden_dim))
Mlp = property(lambda self: Axis(name="mlp", size=self.intermediate_dim))
```

For real examples and deeper understanding, check out `Gpt2Config` in [gpt2.py](https://github.com/stanford-crfm/levanter/blob/main/src/levanter/models/gpt2.py) and `LlamaConfig` in [llama.py](https://github.com/stanford-crfm/levanter/blob/main/src/levanter/models/llama.py).

> To convert your config class to and from Hugging Face config class, you can write class functions `to_hf_config()` and `from_hf_config()` that maps the input parameters to the corresponding Hugging Face parameters. For example, in Llama, we have the following:

Reviewer: probably here talk about HFCompatConfig

```python
from transformers import PretrainedConfig as HfPretrainedConfig

# ...

@classmethod
def from_hf_config(cls, hf_config: HfPretrainedConfig):
    return LlamaConfig(
        seq_len=hf_config.max_position_embeddings,
        hidden_dim=hf_config.hidden_size,
        intermediate_dim=hf_config.intermediate_size,
        num_layers=hf_config.num_hidden_layers,
        num_heads=hf_config.num_attention_heads,
        activation_function=hf_config.hidden_act,
        initializer_range=hf_config.initializer_range,
        layer_norm_epsilon=hf_config.rms_norm_eps,
        rope_scaling=hf_config.rope_scaling,
    )
```

Lastly, you should register your head model's class name as a class property. This step can be deferred until the head class is constructed.

> This class property would make it easier to call your model with the config class. For example:

Reviewer: It's more that it makes it work with the automatic model selection stuff in train_lm and friends

```python
@property
def model_type(cls) -> Type["LlamaLMHeadModel"]:
    return LlamaLMHeadModel
```

### Write Model
After you have defined your config class, you can start writing your model.

> You can follow the breakdown of the model in your Hugging Face implmentation. This would make it easier to validate the implementation through unit tests.

Reviewer: ```suggestion
You can follow the breakdown of the model in your Hugging Face implementation. This would make it easier to validate the implementation through unit tests.
```

For example, in GPT2, we have the following breakdown:
- `Gpt2Mlp`
- `Gpt2Attention`
- `Gpt2Block`: a block of Gpt2Mlp and Gpt2Attention
- `Gpt2Transformer`: a stack of Gpt2Block
- `Gpt2Embeddings`: token and position embeddings
- `Gpt2LMHead`: a complete GPT2 model with embedding, transformer, and LM head.

We follow the same breakdown in the implementation of Llama in Levanter.

#### Note on the Implementation Format
- Each class will have its key layers and components defined as attributes and be initialized with a static method `init()`.

> - Each class will inherit from `StateDictSerializationMixin` from `torch_serialization` and Equinox's `Module` class.

Reviewer: classes don't need to inherit `StateDictSerializationMixin` unless they have custom logic

Author: What're the examples of the custom logic?

Author: Oh I found it below

> ### Mapping weights from/to Hugging Face
> To load weights from Hugging Face, you will need to write a class function `from_hf_state_dict()` in each of your model class. It takes in a Hugging Face state_dict and returns a Levanter state_dict.
> 
> For implementation, there are a few helper classes from `torch_serialization` that you can use:
> - To add specific prefix to the keys of Hugging Face state_dict, you can use the helper function `apply_prefix()`. The prefix comes from the name of attributes defined at the beginning of your model class.
> - To unflatten the linear layers of Hugging Face, you can use the helper function `unflatten_linear_params()`.
> - To unstack the transformer blocks of Hugging Face, you can use the helper function `unstack_transformer_blocks()`.

Reviewer: this isn't really right: `Gpt2MLP` doesn't need custom serialization logic.

```suggestion
### Serialization to/from State Dicts

PyTorch and Hugging Face Transformers use "state dicts" as their preferred serialization format, either as pickles or as the new [safetensors](https://github.com/huggingface/safetensors) format. A state dict is a Python `dict` with string keys and tensor values. The keys of the dict are json-ish "key paths" like `model.blocks.0.mlp.c_proj` and the values are the corresponding parameters for that key path.

You can read more about [PyTorch State Dicts here](https://pytorch.org/tutorials/recipes/recipes/what_is_state_dict.html).

Levanter has machinery for (de)serializing to and from state dicts. Simple cases are handled automatically, but often times custom logic is needed.

#### Easy Case: Identical Module Structure

If your module has exactly the same fields with the same names and same shapes as the Hugging Face state dict (e.g. Gpt2Mlp), you don't need to do anything.

#### Medium Case: Different Names

If for some reason you want to use different names from the HF implementation (e.g. because the names from HF aren't clear...), you can implement `StateDictSerializationMixin` and use `_state_dict_key_map` to rename keys. For instance, the `Gpt2Transformer` class has this method:

```python
class Gpt2Transformer(StateDictSerializationMixin, eqx.Module):

    ...
    
    def _state_dict_key_map(self) -> Dict[str, Optional[str]]:
        return {"blocks": "h"}
```

This says that the field called `blocks` in this class should be (de)serialized as `h`, because the HF GPT-2 implementation uses `h`, which in our opinion isn't very clear. You can also "flatten" the submodules of a field by using `None` or even include `.`s in the name if needed.

#### Hard Case: Custom Serialization

If your modules need special logic, you'll need to implement class `StateDictSerializationMixin` and override `to_state_dict` and `from_state_dict`. It takes in a Hugging Face state dict and returns a Levanter state_dict.

For implementation, there are a few helper classes from `torch_serialization` that you can use:
- To add specific prefix to the keys of a state dict, you can use the helper function `apply_prefix()`. The prefix comes from the name of attributes defined at the beginning of your model class.
- To unflatten the linear layers of a state dict, you can use the helper function `unflatten_linear_params()`.
- To unstack the transformer blocks of a state dict, you can use the helper function `unstack_transformer_blocks()`.
```

> For example, below is the implementation of `from_hf_state_dict()` in LlamaAttention:

Reviewer: ```suggestion
For example, below is the implementation of `from_state_dict()` in `LlamaAttention`. `LLamaAttention`, like most attention layers in Levanter projects the embeddings (with shape `(Pos, Embed)`) to a tensor of shape `(Pos, Head, HeadSize)` rather than the convention in Hugging Face Transformers to use `(Pos, Head * HeadSize)` and then reshaping (because PyTorch Linear doesn't support multiple input or output dimensions).

This difference in convention means that we have to flatten linear layers when converting to a Hugging Face Transformers-compatible state dict and unflatten them when they are read in.
```

```python
def from_state_dict(self, state_dict: StateDict, prefix: Optional[str] = None) -> "LlamaAttention":
    # unflatten the linear layers of HF state_dict to match the shape of LlamaAttention
    d = {}
    d.update(unflatten_linear_layers(apply_prefix(prefix, "q_proj"), state_dict, self.q_proj, True))
    d.update(unflatten_linear_layers(apply_prefix(prefix, "k_proj"), state_dict, self.k_proj, True))
    d.update(unflatten_linear_layers(apply_prefix(prefix, "v_proj"), state_dict, self.v_proj, True))
    d.update(unflatten_linear_layers(apply_prefix(prefix, "o_proj"), state_dict, self.o_proj, True))

    return super().from_state_dict(d, prefix)
```

> Similarly, to save weights to Hugging Face, you will need to write a class function `to_hf_state_dict()` in each of your model class.

Reviewer: ```suggestion
Similarly, to save weights to Hugging Face, you will need to write a class function `to_state_dict()` in each of your model class.
```

The correctness of your implementation can be validated through serialization tests, which will be discussed in the next section.

## Write Tests
There are two types of tests that you should write for your model: unit tests and serialization tests.

### Unit Tests
Unit tests are very useful for testing the correctness of your model implementation.

> It is recommended to have at least one test for each of your modules, so that you can make sure that each module is working as expected and capture any surprises early on, before you test them end-to-end.

Reviewer: I honestly don't think this is necessary but it's useful to do if you get stuck to minimize.

In a unit test, you test your module in the following aspects:
1. The forward pass is successful.
2. The output shape is correct.
3. The output value is correct.

> The point 3 is the most important one, but it is also the most difficult one to implement. If you have a reference implementation in another framework, like HuggingFace, you can use it to test your implementation on the output consistency.

Reviewer: ```suggestion
The point 3 is the most important one, but it is also the most difficult one to implement. If you have a reference implementation in another framework, like Hugging Face Transformers, you can use it to test your implementation on the output consistency.
```

> Here is a piece of example code to test a same module of Levanter against HuggingFace:

Reviewer: ```suggestion
Here is a piece of example code to test a same module of Levanter against Hugging Face Transformers:
```

```python
x = random.normal(key, (1, seq_len))
x_torch = torch.from_numpy(np.array(x))  # convert to torch tensor

levanter_rope = LlamaRotaryEmbedding(HeadSize=HeadSize, Pos=Pos)
levanter_output = levanter_rope(seq_len=seq_len)
hf_rope = HFLlamaRotaryEmbedding(dim=hidden_dim, max_position_embeddings=seq_len, device="cpu")
hf_output = hf_rope(x_torch, seq_len=seq_len)

# compare the output values
for jax_out, torch_out in zip(levanter_output, hf_output):
    torch_out = torch_out.numpy()
    assert np.isclose(torch_out, np.array(jax_out.array), rtol=1e-4, atol=1e-4).all(), f"{torch_out} != {jax_out}"
```

> 

Reviewer: ```suggestion

Be sure to use `np.isclose` with reasonably loose tolerances to account for numerical differences.

```

```python
converter = LlamaConfig.default_hf_checkpoint_converter

# initialize the model in HF...
hf_model = transformer.AutoModelForCausalLM(...)

# save the model weight to disk
with tempfile.TemporaryDirectory() as tmpdir:
    ck_path = f"{tmpdir}/hf_model"
    hf_model.save_pretrained(ck_path)

    model = converter.load_pretrained(
        LlamaLMHeadModel,
        ck_path,
        resize_vocab_to_match_tokenizer=False
    )

# compare the output values between Levanter and HF
# ...
```

The serialization tests are very useful for testing the correctness of your implementation and make sure you can load your pretrained HuggingFace model into Levanter.

## Training
After you have implemented your model and validated it through tests, you can start training your model with Levanter.

### Write Training Configuration
To launch a training job, you will need to write a training configuration file in yaml. It includes the dataset, model, and trainer specifications for the training job. You can find many examples in [configs/](https://github.com/stanford-crfm/levanter/tree/main/config).

> Under the model section, you will need to specify the model name as `type` and modify the hyperparameters that you would like to change. For parameters that are not specified, the default values will be used.

Reviewer: the `type` is needed if registered as an `LmConfig`

  num_layers: 2
```

For more details on the training configuration, please refer to [Configuration Guide](./Configuration-Guide.md).

### Launch Training Job
Once you have your training configuration ready and your training environment set up, you can launch a training job with the following command:

```bash
# HUGGING_FACE_HUB_TOKEN is only needed if you want to load private checkpoint from Hugging Face
WANDB_API_KEY=$WANDB_API_KEY \
HUGGING_FACE_HUB_TOKEN=$HUGGING_FACE_HUB_TOKEN \
python levanter/src/levanter/main/train_lm.py --config_path $CONFIG_PATH
```

Check out [Training on Your Own Data](./Training-On-Your-Data.md) for more detailed guide on how to spin off a training cluster and launch a training job.

### Profile Your Model
If you are interested in profiling the training throughput of your model, good news is that it comes for free with automatic job monitoring in Levanter, powered through Weights & Biases.

Once you run a training job, on the corresponding job page on Weights & Biases, you will be able to find a section named "Throughput". It reports metrics like `examples_per_second` and `tokens_per_second` across the training time.

## Tips for Optimization

> 1. Avoid upcasting to float32. Levanter uses float16 by default, which is more memory efficient and faster for training. You should avoid upcasting to float32 unless it is necessary.

Reviewer: ```suggestion
1. Avoid upcasting to float32. Levanter uses bfloat16 by default, which is more memory efficient and faster for training. You should avoid upcasting to float32 unless it is necessary for stability or accuracy.
```
